### ISSUE TITLE
[6.1.0] Cannot complete OAuth process. Could not find an OAuth cookie for shop url

### ISSUE THREAD

**Issue by kdeng** [Shopify/shopify-api-js#686]
# Issue summary

I am trying to demonstrate how to retrieve the Shopify store access_token using the following code, and I keep receiving this error. I reviewed another issue (https://github.com/Shopify/shopify-api-js/issues/582), and it mentioned the issue should be fixed. Unfortunately, I still experience this issue.

```typescript
  shopifyAPI.auth.callback({
      isOnline: false,
      rawResponse: request,
      rawRequest: response
  })
```

## Expected behavior

to have Shopify.auth.callback({ isOnline: true, rawRequest: req, rawResponse: res, }) to not throw an error

```
  const shopifyAPI = shopifyApi({
    apiKey: apiKey,
    apiSecretKey: apiSecret,
    scopes: getShopifyAppScopes(),
    hostName: getShopifyCallbackHostname(),
    hostScheme: SHOPIFY_APP_HOST_SCHEME,
    apiVersion: SHOPIFY_APP_API_VERSION,
    isEmbeddedApp: true,
  })
  const callbackResponse = await shopifyAPI.auth.callback({
    isOnline: false,
    rawResponse: {},
    rawRequest: {}
  })

```

## Actual behavior

it throw the error : 

```
Error: Cannot complete OAuth process. Could not find an OAuth cookie for shop url: 
    at CookieNotFound.ShopifyError [as constructor] (/Volumes/work/workspaces/my-works/testing/node_modules/@shopify/shopify-api/lib/error.ts:4:3)
    at new CookieNotFound (/Volumes/work/workspaces/my-works/testing/node_modules/@shopify/shopify-api/lib/error.js:174:42)
    at Object.<anonymous> (/Volumes/work/workspaces/my-works/testing/node_modules/@shopify/shopify-api/lib/auth/oauth/oauth.ts:139:13)
    at step (/Volumes/work/workspaces/my-works/testing/node_modules/tslib/tslib.js:144:27)
    at Object.next (/Volumes/work/workspaces/my-works/testing/node_modules/tslib/tslib.js:125:57)
    at fulfilled (/Volumes/work/workspaces/my-works/testing/node_modules/tslib/tslib.js:115:62)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Could someone help to take a look at this issue? Thanks.

**Comment by mkevinosullivan** [comment]
Thanks for flagging this ... we need to document this a bit better.  It's recommended to put the call to `shopify.auth.callback` inside a `try...catch` and redirect to OAuth if the the `CookieNotFound` error is thrown.

An example of this can be seen in an older non-Express version of the app template, [here](https://github.com/Shopify/shopify-app-template-node/blob/cli_three_api_six/web/middleware/auth.js#L92-L95).

**Comment by Copdate-Copdate** [comment]
@mkevinosullivan I did like you suggest and it's always going inside a loop:

```
await shopify.auth.begin({
  callbackPath: "/auth/callback",
  isOnline: false,
  rawRequest: req,
  rawResponse: res,
  shop: req.query.shop
});
```

[shopify-api/INFO] Beginning OAuth | {shop: xxxxxx.myshopify.com, isOnline: false, callbackPath: /auth/callback}

[shopify-api/DEBUG] OAuth started, redirecting to https://xxxxxx.myshopify.com/admin/oauth/authorize?client_id=xxxxxx&scope=read_orders%2Cread_products&redirect_uri=http%3A%2F%2Fxxxxxx.ngrok.io%2Fauth%2Fcallback&state=656248294749758&grant_options%5B%5D= | {shop: xxxxxx.myshopify.com, isOnline: false}

```
const callback = await shopify.auth.callback({
  rawRequest: req,
  rawResponse: res
});
```

[shopify-api/INFO] Completing OAuth | {shop: xxxxxx.myshopify.com}
[shopify-api/ERROR] Could not find OAuth cookie | {shop: xxxxxx.myshopify.com}
CookieNotFound [Error]: Cannot complete OAuth process. Could not find an OAuth cookie for shop url: xxxxxx.myshopify.com
    at CookieNotFound.ShopifyError

---

I don't see any cookie in the request header passed to the callback request.

So unless I'm missing something, we cannot execute the callback because of the cookie issue, and redirecting to the OAuth begin process doesn't seem to fix anything.

**Comment by Copdate-Copdate** [comment]
@mkevinosullivan I dig a little more and see that the call to /auth, which execute the "shopify.auth.begin" function return this:

```
Set-Cookie | shopify_app_state=xxxxx;sameSite=lax; secure=true; path=/auth/callback;expires=Thu, 26 Jan 2023 13:48:25 GMT

Set-Cookie | shopify_app_state.sig=xxxxx=;sameSite=lax; secure=true; path=/auth/callback;expires=Thu, 26 Jan 2023 13:48:25 GMT
```

Then, when shopify call the callback auth/callback, there no cookie in both the request & the response.

**Comment by Copdate-Copdate** [comment]
@mkevinosullivan - Currently testing and re-creating the OAuth process without the library. I don't know if it's strict rule, but I read in a few place that signed cookie start with "s:", which isn't the case with the library. The library instead rename the cookie name "shopify_app_state.sig".

Maybe it can help.

**Comment by Copdate-Copdate** [comment]
@mkevinosullivan - Ok, i confirm that the issue is based on how this library is setting up the signed cookie. 

I followed the tutorial at https://shopify.dev/apps/auth/oauth/getting-started#ask-for-permission and it works perfectly.

**Comment by mkevinosullivan** [comment]
One additional thing to try ... add the following

```js
  logger: {
    level: LogSeverity.Debug,  // import { LogSeverity } from "@shopify/shopify-api";
  },
```

to the configuration you pass into the call to `shopifyApi` ... the debug level _may_ reveal a bit more

**Comment by weeksie** [comment]
`LogSeverity.Debug` does not help at all. 

```
[shopify-api/INFO] Completing OAuth | {shop: XXX.myshopify.com}
[shopify-api/ERROR] Could not find OAuth cookie | {shop: XXX.myshopify.com}
```

Going to dig in further, but this is extremely annoying.

**Comment by weeksie** [comment]
Anyway, I was able to get things working by implementing the oauth code manually. 

```typescript
interface AuthConfig {
  apiKey: string;
  secret: string;
  scopes: string[];
  callbackPath: string;
  homePath: string;
  onAuthStart: (shop: string) => void | Promise<void>;
  onAuthSuccess: (response: unknown) => void | Promise<void>;
}

export const shopify = ({
  apiKey,
  secret,
  callbackPath,
  scopes,
  onAuthStart,
  homePath,
  onAuthSuccess
}: AuthConfig) => {
  const verify = verifyWithKey(secret);

  return {
    async auth(ctx: Koa.Context, next: Koa.Next) {
      const params    = ctx.URL.searchParams;
      const shop      = params.get('shop');
      const timestamp = params.get('timestamp');
      const hmac      = params.get('hmac');

      if (!shop) {
        ctx.throw(422, 'missing shop');
      }

      if (!timestamp) {
        ctx.throw(422, 'missing timestamp');
      }

      if (!hmac) {
        ctx.throw(422, 'missing hmac');
      }

      params.delete('hmac');
      const query = params.toString();

      if (!verify(query, hmac)) {
        ctx.throw(401, `cant verify query ${query}`);
      }

      await onAuthStart(shop);

      const shopifyUrl = shopifyAuthUrl({
        apiKey,
        scopes,
        callbackPath,
        shop,
        host: ctx.host,
        grantOptions: 'value',
      });

      ctx.redirect(shopifyUrl);
      await next();
    },

    async callback(ctx: Koa.Context, next: Koa.Next) {
      const params = ctx.URL.searchParams;
      const shop   = params.get('shop');
      const host   = params.get('host');
      const hmac   = params.get('hmac');
      const code   = params.get('code');
      const grant  = params.get('state');

      if (typeof hmac !== 'string') {
        ctx.throw(422, 'bad hmac');
      }
      if (!shop || !shop.match(/[a-zA-Z0-9][a-zA-Z0-9\-]*\.myshopify\.com/)) {
        ctx.throw(422, 'bad shop');
      }

      if (typeof code !== 'string') {
        ctx.throw(422, 'bad code');
      }

      if (typeof host !== 'string') {
        ctx.throw(422, 'bad host')
      }

      params.delete('hmac');
      const query = params.toString();

      if (!verify(query, hmac)) {
        ctx.throw(401);
      }

      const tokenParams = new URLSearchParams();

      tokenParams.set('client_id', apiKey);
      tokenParams.set('client_secret', secret);
      tokenParams.set('code', code);

      const tokenResponse = await fetch(
        `https://${shop}/admin/oauth/access_token`,
        {
          method: 'post',
          body: tokenParams,
        }
      );

      if (!tokenResponse.ok) {
        console.error(tokenResponse.status);
        console.error(await tokenResponse.text());
        ctx.throw(401);
      }

      const response = camelKeys(await tokenResponse.json());

      await onAuthSuccess(response);

      if (grant === 'value') {
        const shopifyUrl = shopifyAuthUrl({
          apiKey,
          shop,
          scopes,
          callbackPath,
          host: ctx.host,
          grantOptions: 'per-user',
        })

        ctx.redirect(shopifyUrl);
        await next();
        return;
      }


      const isEmbedded = params.get('embedded') === "1";

      if (isEmbedded) {
        const shopifyHost = Buffer.from(host, 'base64').toString();
        const redirect    = `https://${shopifyHost}/apps/${apiKey}?host=${host}&shop=${shop}`;

        ctx.redirect(redirect);
        await next();
        return;
      }

      ctx.redirect(`/${homePath}?shop=${shop}&host=${encodeURIComponent(host)}`);
      await next();
      return;
    }
  }
};

type UrlProps = {
  apiKey: string;
  grantOptions: 'per-user' | 'value';
  callbackPath: string;
  shop: string;
  host: string;
  scopes: string[];
}

const shopifyAuthUrl = (props: UrlProps) => {
  const shopifyParams = new URLSearchParams();

  shopifyParams.set('client_id', props.apiKey);
  shopifyParams.set('scope', props.scopes.join(','));
  shopifyParams.set('redirect_uri', `https://${props.host}/${props.callbackPath}`)
  shopifyParams.set('grant_options[]', props.grantOptions);
  shopifyParams.set('state', props.grantOptions);

  return `https://${props.shop}/admin/oauth/authorize?${shopifyParams}`;
};
```

The only thing missing is the code for `camelKeys` and `verifyWithKey` but that shouldn't be the problem. I'm not setting cookies anywhere though. Are they strictly necessary? What are you using cookies for, and or should I be setting them?

**Comment by bhr** [comment]
Hitting the same issue, but the fix to catch the CookieNotFoundError and try again isn't helping. It's just not setting the cookie

**Comment by zds97** [comment]
I'm also hitting this same issue.  Restarting the auth process hen the error is an instanceof CookieNotFound also caused an infinite redirect.

**Comment by mimiminimalizm** [comment]
We were able to reproduce the issue if the shop domain is not standard (like XXXXX.myshopify.com), but the custom one. Because, yeah, redirecting to the OAuth if the the CookieNotFound error found not helping if the cookie even wasn't set

**Comment by The1987** [comment]
could you also get around the issue by doing the following?

- if shop exists then look for session in storage
    if expired push through oauth and save new session
    if not expired then push to app frontend

- if !shop then run through oauth

- then whenever you make a call you can check to see if the session is or isnt valid then perform following steps...

**Comment by The1987** [comment]
@zds97 - Thanks for your feedback. Again, what I was seeing was that on the first install the session did generate (using Chrome), however it wouldn't work afterwards, and thus I got the error.

Originally I implemented what I suggested, but then I remember that the [isOnline allows for true or false](https://shopify.dev/docs/apps/auth/oauth/access-modes). My app doesn't require such strict session requirements so I have set it to false, and thus far working as intended on Chrome.

I haven't tested, but I wonder if having the the isOnline set to false and trying to "re-oauth" will throw the error vs having it set to true may not throw the error. Just taking a stab in the dark here. Again, I would also hypothesize that once the session has been created the browser might be storing the session (unbeknown to us) and maybe clearing your browsing data might fix this in terms of unblocking your development.

**Comment by zds97** [comment]
@abharvey Sure.  All of our instances have shops, along with the rest of the Shopify install information associated with the request, so it doesn't seem like Google crawler is causing the issue.  It seems to just be missing the cookie as mentioned above.   

Here's an example from yesterday and the userAgent logging:

userAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36

I can add additional logging if there's something specific you're looking for.

**Comment by abharvey** [comment]
@zds97  Thanks for getting back to us quickly. We're having difficulty reproducing the error consistently with the exception of when Chrome tries to crawl the routes. As far as logs that could help, if you could provide some debug logs before and after the error you mentioned in the issue that may help.
If you're able to reproduce consistently perhaps an example app repo could help us debug as well. 

@flamendless Could you provide any errors you're seeing in the server side logs?

**Comment by flamendless** [comment]
@zds97 i dont get detailed logs (idky how to configure the loggers) but heres what i get same as others: (no access to pc right now so I'll just copy-paste from the other comments)

Error: Cannot complete OAuth process. Could not find an OAuth cookie for shop url: 

And the

[shopify-api/ERROR] Could not find OAuth cookie | {shop: XXX.myshopify.com}

Im using the template provided by the shopify cli using shopify/shopify-express-app

**Comment by abharvey** [comment]
@flamendless Thanks, since you're using the app-express template you can edit your config to add addition logging like this:

// in shopify.js (unless renamed)
```js
// add import for the enum
import { LogSeverity } from "@shopify/shopify-api";

const shopify = shopifyApp({
  api: {
    logger: {
      level: LogSeverity.Debug,
    }
//... Rest of your config
```

**Comment by flamendless** [comment]
@abharvey @zds97 Just tried again, now it works. I can install the app to my  dev store properly. Weird

**Comment by cmelendez** [comment]
Some people might find [this] (https://github.com/Shopify/shopify-api-js/issues/726#issuecomment-1518228755) useful.

**Comment by anhnt2310** [comment]
Any updates or solutions for this issue? It's an emergency because of affecting old apps upgrading to catch up with the latest API. so in case this issue couldn't be resolved soon, the app couldn't be used, affecting to revenue of the business which develops the app on Shopify.

**Comment by cmelendez** [comment]
I've created a [PR](https://github.com/Shopify/shopify-api-js/pull/905) that, at least for our case, solves the issue by allowing to set an optional cookie domain (by default the library uses the domain that starts the oauth flow).

**Comment by kdichev** [comment]
Experiencing the same issue with oauth flow when im routing the requests through a few services on different domains, hopefully @cmelendez PR gets in soon!

**Comment by The1987** [comment]
> 

@wweksie for the WIN!

I want to confirm that changing the samesite within the Shopify oauth package from lax to none does indeed work to resolve the issue on Chrome.

Here is a screenshot of where this package live:
<img width="387" alt="Screenshot 2023-07-21 at 5 25 05 PM" src="https://github.com/Shopify/shopify-api-js/assets/28612269/9e4f2968-bd20-41ff-b92d-bc35bf7a8731">

**Comment by jsegeren** [comment]
We have the same issue - sometimes getting "Could not find OAuth cookie" error, but not consistently. 
Right now, this is blocking us from publishing a new app to the Shopify App Store, since the reviewers are rejecting it (!!)

I previously posted in what looks like a duplicate thread - https://github.com/Shopify/shopify-api-js/issues/933

We'd REALLY appreciate an update from the Shopify/core team on this issue. Exploring workarounds in the interim.

As you can see from this set of example logs, for the same shop, same setup, etc. back to back OAuth attempts. The first time fails, the second time succeeds.

[oauth-logs.log](https://github.com/Shopify/shopify-api-js/files/12325944/oauth-logs.log)

**Comment by Sourabhtripathi1** [comment]
> 

Can you guide me on How you did that, I am suffering from the same problem.
It works properly on the development server but it can not find the cookie while I run its production version on the server

**Comment by Andrew-Golchenko** [comment]
have the same issue :(((
Error during OAuth callback | {error: Cannot complete
OAuth process. Could not find an OAuth cookie for shop url: **-test.myshopify.com}

**Comment by axcasella** [comment]
> Catching the CookieNotFound error and then redirecting to Shopify again (and back) solves the problem for me. Found it in an example.
> 
> ```ts
>     try {
>       const session = (await this._shopifyAPI.auth.callback({
>         rawRequest: params.request,
>         rawResponse: params.response,
>       })).session;
> 
>       return session?.toObject();
>     } catch (e) {
>       if (e instanceof InvalidOAuthError) {
>         throw new BadRequestException();
>       } else if (e instanceof CookieNotFound) {
>         await this._shopifyAPI.auth.begin({
>           shop: params.shop,
>           isOnline: false,
>           callbackPath: params.callback,
>           rawRequest: params.request,
>           rawResponse: params.response,
>         });
> 
>         return undefined;
>       } else {
>         throw e;
>       }
>     }
> ```

Has anyone managed to find a fix? Tried this and doesn't work. Still getting same error.

**Comment by randidev** [comment]
Just ran into this problem, does anyone have a validated way to solve this? 
I've got a non-embedded apps and here's what I got from the log

```
[shopify-api/INFO] Beginning OAuth | {shop: *.myshopify.com, isOnline: false, callbackPath: /api/v1/shopify/callback}
[shopify-api/INFO] Completing OAuth | {shop: *.myshopify.com}
[shopify-api/ERROR] Could not find OAuth cookie | {shop: *.myshopify.com}
```
